File: java/FunctionalList.java
import java.util.NoSuchElementException;

public abstract class FunctionalList<T> {
    public static <T> FunctionalList<T> nil() {
        return new Nil<T>();
    }

    public static <T> FunctionalList<T> cons(final T head, final FunctionalList<T> tail) {
        return new Cons<T>(head, tail);
    }

    public static <T> FunctionalList<T> of(final T... items) {
        return fromArray(items, 0, items.length);
    }

    private static <T> FunctionalList<T> fromArray(final T[] items, final int start, final int end) {
        return start == end
                   ? FunctionalList.<T>nil()
                   : cons(items[start], fromArray(items, start + 1, end));
    }

    public abstract boolean isEmpty();

    public abstract T head();

    public abstract FunctionalList<T> tail();

    public abstract <U> FunctionalList<U> map(Function<T, U> mapping);

    public static final class Nil<T> extends FunctionalList<T> {
        @Override
        public boolean isEmpty() {
            return true;
        }

        @Override
        public T head() {
            throw new NoSuchElementException();
        }

        @Override
        public FunctionalList<T> tail() {
            throw new NoSuchElementException();
        }

        @Override
        public <U> FunctionalList<U> map(Function<T, U> mapping) {
            return this;
        }
    }

    public static final class Cons<T> extends FunctionalList<T> {
        private final T head;
        private final FunctionalList<T> tail;

        public Cons(final T head, final FunctionalList<T> tail) {
            this.head = head;
            this.tail = tail;
        }

        @Override
        public boolean isEmpty() {
            return false;
        }

        @Override
        public T head() {
            return head;
        }

        @Override
        public FunctionalList<T> tail() {
            return tail;
        }

        @Override
        public <U> FunctionalList<U> map(Function<T, U> mapping) {
            return cons(mapping.apply(head()), tail().map(mapping));
        }
    }
}

File: csharp/FunctionalList.cs
using System;

namespace DysfunctionalProgramming
{
    public abstract class FunctionalList<T> where T: IComparable
    {
        public static FunctionalList<T> Nil()
        {
            return new Nil<T>();
        }

        public static FunctionalList<T> Cons(T head, FunctionalList<T> tail)
        {
            return new Cons<T>(head, tail);
        }

        public static FunctionalList<T> Of(params T[] items)
        {
            return FromArray(items, 0, items.Length);
        }

        private static FunctionalList<T> FromArray(T[] items, int start, int end)
        {
            return start == end
                ? Nil()
                    : Cons(items[start], FromArray(items, start + 1, end));
        }

        public abstract bool IsEmpty { get; }

        public abstract T Head { get; }

        public abstract FunctionalList<T> Tail { get; }

        public abstract FunctionalList<TResult> Map<TResult>(Func<T, TResult> mapping) where TResult: IComparable;
    }

    sealed class Nil<T> : FunctionalList<T> where T: IComparable
    {
        public override bool IsEmpty {
            get { return true; }
        }

        public override T Head {
            get { throw new InvalidOperationException(); }
        }

        public override FunctionalList<T> Tail {
            get { throw new InvalidOperationException(); }
        }

        public override FunctionalList<TResult> Map<TResult>(Func<T, TResult> mapping)
        {
            return new Nil<TResult>();
        }
    }

    sealed class Cons<T> : FunctionalList<T> where T: IComparable
    {
        private readonly T _head;
        private readonly FunctionalList<T> _tail;

        public Cons(T head, FunctionalList<T> tail)
        {
            _head = head;
            _tail = tail;
        }

        public override bool IsEmpty {
            get { return false; }
        }

        public override T Head {
            get { return _head; }
        }

        public override FunctionalList<T> Tail {
            get { return _tail; }
        }

        public override FunctionalList<TResult> Map<TResult>(Func<T, TResult> mapping)
        {
            return FunctionalList<TResult>.Cons(mapping(Head), Tail.Map(mapping));
        }
    }
}

